chain-impl-mockchain/src/evm/mod.rs

```diff
@@ -126,152 +191,39 @@ pub fn serialize_u256<T>(bb: ByteBuilder<T>, value: &U256) -> ByteBuilder<T> {
-            let keys_count = codec.get_be_u64().unwrap_or_default();
-            let keys = (0..keys_count).into_iter().fold(Vec::new(), |mut keys, _| {
-                let key = read_h256(codec).unwrap_or_default();
-                if !key.is_zero() {
-                    keys.push(key);
-                }
-                keys
-            });
-            access_list.push((address, keys));
-            access_list
-        });
-    Ok(access_list)
-}
-
-impl DeserializeFromSlice for EvmTransaction {
-    fn deserialize_from_slice(codec: &mut Codec<&[u8]>) -> Result<Self, ReadError> {
-        let contract_type = codec.get_u8()?;
-        match contract_type {
-            #[cfg(feature = "evm")]
-            0 => {
-                // CREATE Transaction
-                let caller = read_address(codec)?;
-                let value = read_u256(codec)?;
-                let init_code = read_bytecode(codec)?;
-                let gas_limit = read_gas_limit(codec)?;
-                let access_list = read_access_list(codec)?;
-                Ok(EvmTransaction::Create {
-                    caller,
-                    value,
-                    init_code,
-                    gas_limit,
-                    access_list,
-                })
-            }
-            #[cfg(feature = "evm")]
-            1 => {
-                // CREATE2 Transaction
-                let caller = read_address(codec)?;
-                let value = read_u256(codec)?;
-                let init_code = read_bytecode(codec)?;
-                let salt = read_h256(codec)?;
-                let gas_limit = read_gas_limit(codec)?;
-                let access_list = read_access_list(codec)?;
-                Ok(EvmTransaction::Create2 {
-                    caller,
-                    value,
-                    init_code,
-                    salt,
-                    gas_limit,
-                    access_list,
-                })
-            }
-            #[cfg(feature = "evm")]
-            2 => {
-                // CALL Transaction
-                let caller = read_address(codec)?;
-                let address = read_address(codec)?;
-                let value = read_u256(codec)?;
-                let data = read_bytecode(codec)?;
-                let gas_limit = read_gas_limit(codec)?;
-                let access_list = read_access_list(codec)?;
-                Ok(EvmTransaction::Call {
-                    caller,
-                    address,
-                    value,
-                    data,
-                    gas_limit,
-                    access_list,
-                })
-            }
-            n => Err(ReadError::UnknownTag(n as u32)),
+impl Deserialize for EvmTransaction {
+    fn deserialize<R: std::io::Read>(_codec: &mut Codec<R>) -> Result<Self, ReadError> {
+        #[cfg(feature = "evm")]
+        {
+            let mut rlp_bytes = vec![];
+            _codec.read_to_end(&mut rlp_bytes)?;
```

Comment:
`read_to_end` is what causes the deser errors, since the rest of the transaction fields are _after_ the payload, and this is consuming everything. So the expiry date is not really invalid, it's just that there isn't anything else to read on the buffer.

```suggestion
            let len = _codec.get_be_u64()?;
            let rlp_bytes = _codec.get_bytes(len as usize)?;
```